### PR TITLE
Convert membership date fields to datepicker & form to entity form

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
As part of our ongoing effort to convert all date fields to datepicker field & a newer interest in cleaning up the membership form this converts 3 date fields on the Membership form. It uses the new & preferred entityFormTrait to add the fields

Before
----------------------------------------
Old style date fields

After
----------------------------------------
New style date fields

Technical Details
----------------------------------------
Removing jcalendar has been a pretty long-running initiative - although we still have 68 instances even after this :-( The new datepicker format requires less date wrangling & hence is more robust & allows simplification. Ditto the EntityFormTrait allows us to move to a more metadata oriented approach

Comments
----------------------------------------
@mattwire @seamuslee001 @colemanw

### CODE REVIEW

File: CRM/Core/Form/EntityFormTrait.php [hunk 1]
```diff
@@ -202,12 +202,23 @@ protected function setTranslatedFields() {
   protected function addEntityFieldsToTemplate() {
     foreach ($this->getEntityFields() as $fieldSpec) {
       if (empty($fieldSpec['not-auto-addable'])) {
-        $element = $this->addField($fieldSpec['name'], [], CRM_Utils_Array::value('required', $fieldSpec));
+        $element = $this->addField($fieldSpec['name'], [], CRM_Utils_Array::value('required', $fieldSpec), FALSE);
```
Comment (eileenmcnaughton) [inline CRM/Core/Form/EntityFormTrait.php]: The FALSE here denotes 'dont' use legacy date field type' - we don't have date fields in other entityTrait forms hence this wasn't set earlier

File: CRM/Member/Form/Membership.php [hunk 2]
```diff
@@ -112,14 +112,54 @@ class CRM_Member_Form_Membership extends CRM_Member_Form {
...
+    ];
+  }
+
+  /**
+   * Set the delete message.
    *
-   * @return int|null
+   * We do this from the constructor in order to do a translation.
    */
-  public function getEntityId() {
```
Comment (eileenmcnaughton) [inline CRM/Member/Form/Membership.php]: trait has this function

File: CRM/Member/Form/Membership.php [hunk 3]
```diff
@@ -562,19 +599,10 @@ public function buildQuickForm() {
       $sel->freeze();
     }
 
-    $this->applyFilter('__ALL__', 'trim');
```
Comment (eileenmcnaughton) [inline CRM/Member/Form/Membership.php]: buildQuickEntityForm does this

File: CRM/Core/Form/EntityFormTrait.php [hunk 4]
```diff
@@ -202,12 +202,23 @@ protected function setTranslatedFields() {
...
   }
 
+  /**
+   * Is the form being used in the context of a deletion.
+   *
+   * (For some reason rather than having separate forms Civi overloads one form).
+   *
+   * @return bool
+   */
+  protected function isDeleteContext() {
```
Comment (eileenmcnaughton) [inline CRM/Core/Form/EntityFormTrait.php]: I didn't turn out to need this extracted fn - but still - it's nice & clear :-)

File: CRM/Member/Form/Membership.php [hunk 5]
```diff
@@ -182,6 +222,9 @@ public function preProcess() {
     // This string makes up part of the class names, differentiating them (not sure why) from the membership fields.
     $this->assign('formClass', 'membership');
     parent::preProcess();
+    if ($this->isUpdateToExistingRecurringMembership()) {
+      $this->entityFields['end_date']['is_freeze'] = TRUE;
```
Comment (eileenmcnaughton) [inline CRM/Member/Form/Membership.php]: We can't do this freeze until preProcess as id is not retrieved from the form earlier. I did go down the path of trying to move the call to setEntityFields to the preProcess - but perhaps it makes sense to keep it generic & tweak it in preProcess - as done here.

There are other fields that will ideally be moved into this & into the setEntityFields

File: CRM/Member/Form/Membership.php [hunk 6]
```diff
@@ -372,16 +415,9 @@ public function setDefaultValues() {
       // $defaults['credit_card_exp_date'] = array( 'Y' => '2012', 'M' => '05' );
     }
 
-    $dates = array('join_date', 'start_date', 'end_date');
-    foreach ($dates as $key) {
-      if (!empty($defaults[$key])) {
-        list($defaults[$key]) = CRM_Utils_Date::setDateDefaults(CRM_Utils_Array::value($key, $defaults));
```
Comment (eileenmcnaughton) [inline CRM/Member/Form/Membership.php]: just yay

File: CRM/Member/Form/Membership.php [hunk 7]
```diff
@@ -1934,9 +1962,9 @@ protected function setStatusMessage($membership, $endDate, $receiptSend, $member
   protected function isUpdateToExistingRecurringMembership() {
     $isRecur = FALSE;
     if ($this->_action & CRM_Core_Action::UPDATE
-      && CRM_Core_DAO::getFieldValue('CRM_Member_DAO_Membership', $this->_id,
+      && CRM_Core_DAO::getFieldValue('CRM_Member_DAO_Membership', $this->getEntityId(),
```
Comment (eileenmcnaughton) [inline CRM/Member/Form/Membership.php]: just for form

**Comment by colemanw** [comment]
@civicrm-builder retest this please

**Comment by eileenmcnaughton** [comment]
@colemanw it passed now

**Comment by mattwire** [comment]
Ok I've given this a quick test and am happy with it.  @colemanw If you are happy please merge.

**Pull request merged**